pkg/cmd/cli/cmd/debug.go

```diff
@@ -0,0 +1,550 @@
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/util/interrupt"
+	"k8s.io/kubernetes/pkg/util/term"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/watch"
+
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type DebugOptions struct {
+	Attach kcmd.AttachOptions
+
+	Print         func(pod *kapi.Pod, w io.Writer) error
+	LogsForObject func(object, options runtime.Object) (*kclient.Request, error)
+
+	NoStdin    bool
+	ForceTTY   bool
+	DisableTTY bool
+	Filename   string
+	Timeout    time.Duration
+
+	Command         []string
+	Annotations     map[string]string
+	AsRoot          bool
+	KeepLabels      bool
+	KeepAnnotations bool
+	KeepLiveness    bool
+	KeepReadiness   bool
+	OneContainer    bool
+	NodeName        string
+	AddEnv          []kapi.EnvVar
+	RemoveEnv       []string
+}
+
+const (
+	debugLong = `
+Launch a command shell to debug a running application
+
+When debugging images and setup problems, it's useful to get an exact copy of a running
+pod configuration and troubleshoot with a shell. Since a pod that is failing may not be
+started and not accessible to 'rsh' or 'exec', the 'debug' command makes it easy to
+create a carbon copy of that setup.
+
+The default mode is to start a shell inside of the first container of the referenced pod,
+replication controller, or deployment config. The started pod will be a copy of your
+source pod, with labels stripped, the command changed to '/bin/sh', and readiness and
+liveness checks disabled. If you just want to run a command, pass -I to disable the
+shell and add '--' and a command to run. Other flags are supported for altering the
+container or pod in common ways.
+
+The debug pod is deleted when the the remote command completes or the user interrupts
+the shell.`
+
+	debugExample = `
+  # Debug a currently running deployment
+  $ %[1]s dc/test
+
+  # Debug a specific failing container by running the env command in the 'second' container
+  $ %[1]s dc/test -c second -- /bin/env
+
+  # See the pod that would be created to debug
+  $ %[1]s dc/test -o yaml`
+
+	debugPodLabelName = "debug.openshift.io/name"
+
+	debugPodAnnotationSourceContainer = "debug.openshift.io/source-container"
+	debugPodAnnotationSourceResource  = "debug.openshift.io/source-resource"
+)
+
+// NewCmdDebug creates a command for debugging pods.
+func NewCmdDebug(fullName string, f *clientcmd.Factory, in io.Reader, out, errout io.Writer) *cobra.Command {
+	options := &DebugOptions{
+		Timeout: 30 * time.Second,
+		Command: []string{"/bin/sh"},
+		Attach: kcmd.AttachOptions{
+			In:    in,
+			Out:   out,
+			Err:   errout,
+			TTY:   true,
+			Stdin: true,
+
+			Attach: &kcmd.DefaultRemoteAttach{},
+		},
+		LogsForObject: f.LogsForObject,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "debug [RESOURCE/]NAME [ENV1=VAL1 ...] [-- COMMAND]",
```

Comment:
Most people won't be able to as root though.  Will add options.

On Thu, Feb 18, 2016 at 11:37 AM, Fabiano Franz <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/debug.go
> <https://github.com/openshift/origin/pull/7294#discussion_r53341125>:
>
> > +		Timeout: 30 * time.Second,
> > +		Command: []string{"/bin/sh"},
> > +		Attach: kcmd.AttachOptions{
> > +			In:    in,
> > +			Out:   out,
> > +			Err:   errout,
> > +			TTY:   true,
> > +			Stdin: true,
> > +
> > +			Attach: &kcmd.DefaultRemoteAttach{},
> > +		},
> > +		LogsForObject: f.LogsForObject,
> > +	}
> > +
> > +	cmd := &cobra.Command{
> > +		Use:     "debug [RESOURCE/]NAME [ENV1=VAL1 ...] [-- COMMAND]",
>
> I'd also add -c and --as-root to the usage, they seem of pretty common
> use. So
>
> debug [RESOURCE/]NAME [ENV1=VAL1 ...] [-c CONTAINER] [--as-root] [-- COMMAND]
>
> Also, note that our templates adds [options] automatically to the end of
> usage, but in this case it would be after --. So you may want to add
> [options] manually right before [-- COMMAND].
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7294/files#r53341125>.
>